README.md
picoGuild is a small [LibDataBroker](http://github.com/tekkub/libdatabroker-1-1)
based info addon. It provides a simple guild display, with a tooltip that shows
detailed information about guildies online.


changelog.txt
6.0.0.13-Beta
    Remove guild level/XP stuff for 6.0 (thanks @hughescr and @Gnarfoz)

5.0.1.12-Release
    Tagging release version

5.0.1.11-Beta
    Guild daily XP cap is gone (thanks hughescr)

4.1.0.10-Beta
    Display party/raid indicator more betterer
    Identify party/raid members (thanks dcollier)

4.0.0.9-Beta
    Seems string.format rounds instead of truncating

4.0.0.8-Beta
    remove wrath backcompat
    pprovost is holding it wrong

4.0.0.7-Beta
    Open the new guild frame on click
    Live only has partial guild functionality right now
    new guild level and rep infos in cat

3.1.0.6-Release
    Add zone highlighting to tooltip

3.0.3.5-Release
    LDB type

3.0.3.4-Beta
    tektip!
    Embed update

3.0.1.3-Beta
    Add location info to tooltip
    Embed update

2.4.3.2-Beta
    Change officer note color so you can tell it apart from the public note
    Use token class name instead of engrish one

2.4.2.1-Beta
    Add LDB-1.1
    Remove tekBlock, we're no longer a display
    Rename to picoGuild
    Remove dongle


tektip.lua

local lib, oldminor = LibStub:NewLibrary("tektip-1.0", 3)
if not lib then return end
oldminor = oldminor or 0


local PADDING, VGAP, HGAP = 5, 2, 5
local TOOLTIP_PADDING = PADDING + HGAP

local bgFrame = {bgFile = "Interface\\Tooltips\\UI-Tooltip-Background", insets = {left = PADDING, right = PADDING, top = PADDING, bottom = PADDING},
	tile = true, tileSize = 16, edgeFile = "Interface\\Tooltips\\UI-Tooltip-Border", edgeSize = 16}

if oldminor < 1 then
	local linemeta = {__index = function(t,i)
		if i == "header" then
			local h = CreateFrame("Frame", nil, t.frame)
			h:SetHeight(1)
			if t.colnum == 1 then
				h:SetPoint("TOPLEFT", t.frame, "TOPLEFT", TOOLTIP_PADDING, -TOOLTIP_PADDING)
			else
				h:SetPoint("LEFT", t.frame.cols[t.colnum-1].header, "RIGHT", HGAP, 0)
			end
			t[i] = h
			return h
		end

		local fs = t.frame:CreateFontString(nil, nil, i == 1 and "GameTooltipHeaderText" or "GameTooltipText")
		if t.colnum == 1 then
			if i == 1 then
				fs:SetPoint("TOPLEFT", t.frame, "TOPLEFT", TOOLTIP_PADDING, -TOOLTIP_PADDING)
			else
				fs:SetPoint("TOPLEFT", t[i-1], "BOTTOMLEFT", 0, -VGAP)
			end
		else
			fs:SetPoint("TOP", t.frame.cols[t.colnum-1][i], "TOP")
			fs:SetPoint("LEFT", t.header, "LEFT")
			fs:SetPoint("RIGHT", t.header, "RIGHT")
		end
		fs:SetJustifyH(t.justify)
		t[i] = fs
		return fs
	end}

	function lib.new(cols, ...)
		assert(type(cols) == "number" and cols >= 2, "Must have at least two columns.")

		local f = CreateFrame("Frame", nil, UIParent)
		f:SetBackdrop(bgFrame)
		f:SetBackdropBorderColor(TOOLTIP_DEFAULT_COLOR.r, TOOLTIP_DEFAULT_COLOR.g, TOOLTIP_DEFAULT_COLOR.b)
		f:SetBackdropColor(TOOLTIP_DEFAULT_BACKGROUND_COLOR.r, TOOLTIP_DEFAULT_BACKGROUND_COLOR.g, TOOLTIP_DEFAULT_BACKGROUND_COLOR.b)
		f:SetFrameStrata("TOOLTIP")
		f:SetClampedToScreen(true)

		f.cols = {}
		for i=1,cols do
			local justify = select(i, ...) or i == 1 and "LEFT" or i == cols and "RIGHT" or "CENTER"
			f.cols[i] = setmetatable({frame = f, colnum = i, justify = justify}, linemeta)
		end

		f.AddLine = lib.AddLine
		f.AddMultiLine = lib.AddMultiLine
		f.Clear = lib.Clear
		f.AnchorTo = lib.AnchorTo

		f:SetScript("OnHide", lib.Clear)
		f:SetScript("OnShow", lib.OnShow)

		f:Hide()
		return f
	end


	local function GetTipAnchor(frame)
		local x,y = frame:GetCenter()
		if not x or not y then return "TOPLEFT", "BOTTOMLEFT" end
		local hhalf = (x > UIParent:GetWidth()*2/3) and "RIGHT" or (x < UIParent:GetWidth()/3) and "LEFT" or ""
		local vhalf = (y > UIParent:GetHeight()/2) and "TOP" or "BOTTOM"
		return vhalf..hhalf, frame, (vhalf == "TOP" and "BOTTOM" or "TOP")..hhalf
	end

	function lib:AnchorTo(frame)
		self:Clear()
		self:ClearAllPoints()
		self:SetPoint(GetTipAnchor(frame))
	end


	function lib:Clear()
		self.numlines = 0
		for i,col in pairs(self.cols) do
			for j,fs in ipairs(col) do
				fs:Hide()
			end
		end
	end
end


if oldminor < 3 then
	function lib:AddLine(text, r, g, b, wrap)
		self.numlines = self.numlines + 1
		local fs = self.cols[1][self.numlines]
		fs:SetTextColor(r or NORMAL_FONT_COLOR.r, g or NORMAL_FONT_COLOR.g, b or NORMAL_FONT_COLOR.b)
		fs:SetText(text)
		fs:SetPoint("RIGHT", self.cols[#self.cols].header, "RIGHT")
		fs.wrapped = wrap
		for i=2,#self.cols do self.cols[i][self.numlines]:SetText(nil) end
		fs:Show()
	end
end


if oldminor < 1 then
	function lib:AddMultiLine(...)
		self.numlines = self.numlines + 1
		local numdots = select("#", ...)
		for i=1,#self.cols do
			local fs = self.cols[i][self.numlines]
			local text = select(i, ...)
			local r, g, b
			if numdots >= (#self.cols + i*3) then r, g, b = select(#self.cols + i*3 - 2, ...) end
			fs:SetTextColor(r or NORMAL_FONT_COLOR.r, g or NORMAL_FONT_COLOR.g, b or NORMAL_FONT_COLOR.b)
			fs:SetText(text == "" and " " or text)
			if i == 1 then fs:SetPoint("RIGHT", self.cols[1].header, "RIGHT") end
			fs:Show()
		end
	end
end


if oldminor < 3 then
	function lib:OnShow()
		local w, totalw = 0, 0
		for i=1,#self.cols do
			local colw = 1
			for j=1,self.numlines do
				local fs = self.cols[i][j]
				if i ~= 1 or self.cols[2][j]:GetText() then colw = math.max(colw, fs:GetStringWidth())
				elseif not fs.wrapped then totalw = math.max(totalw, fs:GetStringWidth()) end
			end
			self.cols[i].header:SetWidth(colw)
			w = w + colw
		end

		if w < (totalw - (#self.cols-1)*HGAP) then
			local extra = totalw - w - (#self.cols-1)*HGAP
			for i=1,#self.cols do self.cols[i].header:SetWidth(extra/#self.cols + self.cols[i].header:GetWidth()) end
			w = totalw
		else
			w = w + (#self.cols-1)*HGAP
		end
		self:SetWidth(w + TOOLTIP_PADDING*2)

		local h = 0
		for i=1,self.numlines do
			local fs = self.cols[1][i]
			fs:SetWidth(fs.wrapped and w or 0)
			h = h + fs:GetHeight()
		end

		self:SetHeight(h + TOOLTIP_PADDING*2 + (self.numlines-1)*VGAP)
	end
end


GuildBlock.lua


----------------------------
--      Localization      --
----------------------------

local L = {
	offline = "(.+) has gone offline.";
	online = "|Hplayer:%s|h[%s]|h has come online.";	["has come online"] = "has come online",
	["has gone offline"] = "has gone offline",

	["No Guild"] = "No Guild",
	["Not in a guild"] = "Not in a guild",
}


------------------------------
--      Are you local?      --
------------------------------

local mejoin = UnitName("player").." has joined the guild."
local friends, colors = {}, {}
for class,color in pairs(RAID_CLASS_COLORS) do colors[class] = string.format("%02x%02x%02x", color.r*255, color.g*255, color.b*255) end


-------------------------------------------
--      Namespace and all that shit      --
-------------------------------------------

local dataobj = LibStub:GetLibrary("LibDataBroker-1.1"):NewDataObject("GuildBlock", {icon = "Interface\\Addons\\GuildBlock\\icon", text = L["No Guild"]})
local f = CreateFrame("Frame")
f:SetScript("OnEvent", function(self, event, ...) if self[event] then return self[event](self, event, ...) end end)


----------------------------------
--      Server query timer      --
----------------------------------

local MINDELAY, DELAY = 15, 300
local elapsed, dirty = 0, false
f:Hide()
f:SetScript("OnUpdate", function(self, elap)
	elapsed = elapsed + elap
	if (dirty and elapsed >= MINDELAY) or elapsed >= DELAY then
		if IsInGuild() then GuildRoster() else elapsed, dirty = 0, false end
	end
end)


local orig = GuildRoster
GuildRoster = function(...)
	elapsed, dirty = 0, false
	return orig(...)
end


---------------------------
--      Init/Enable      --
---------------------------

function f:PLAYER_LOGIN()
	self:Show()
	self:RegisterEvent("GUILD_ROSTER_UPDATE")
	self:RegisterEvent("CHAT_MSG_SYSTEM")

	SortGuildRoster("class")
	if IsInGuild() then GuildRoster() end

	self:UnregisterEvent("PLAYER_LOGIN")
	self.PLAYER_LOGIN = nil
end


------------------------------
--      Event Handlers      --
------------------------------

function f:CHAT_MSG_SYSTEM(event, msg)
	if string.find(msg, L["has come online"]) or string.find(msg, L["has gone offline"]) or msg == mejoin then dirty = true end
end


function f:GUILD_ROSTER_UPDATE()
	local online = 0

	if IsInGuild() then
		for i = 1,GetNumGuildMembers(true) do if select(9, GetGuildRosterInfo(i)) then online = online + 1 end end
		dataobj.text = string.format("%d/%d", online, GetNumGuildMembers(true))
	else dataobj.text = L["No Guild"] end
end


------------------------
--      Tooltip!      --
------------------------

local function GetTipAnchor(frame)
	local x,y = frame:GetCenter()
	if not x or not y then return "TOPLEFT", "BOTTOMLEFT" end
	local hhalf = (x > UIParent:GetWidth()*2/3) and "RIGHT" or (x < UIParent:GetWidth()/3) and "LEFT" or ""
	local vhalf = (y > UIParent:GetHeight()/2) and "TOP" or "BOTTOM"
	return vhalf..hhalf, frame, (vhalf == "TOP" and "BOTTOM" or "TOP")..hhalf
end


function dataobj.OnLeave() GameTooltip:Hide() end
function dataobj.OnEnter(self)
 	GameTooltip:SetOwner(self, "ANCHOR_NONE")
	GameTooltip:SetPoint(GetTipAnchor(self))
	GameTooltip:ClearLines()

	if IsInGuild() then
		GameTooltip:AddDoubleLine("GuildBlock", GetGuildInfo("player"))
		GameTooltip:AddLine(GetGuildRosterMOTD(), 0, 1, 0, true)
		GameTooltip:AddLine(" ")

		local mylevel = UnitLevel("player")
		for i=1,GetNumGuildMembers(true) do
			local name, rank, rankIndex, level, class, area, note, officernote, connected, status = GetGuildRosterInfo(i)
			if connected then
				local levelcolor = (level >= (mylevel - 5) and level <= (mylevel + 5)) and "|cff00ff00" or ""
				GameTooltip:AddDoubleLine(string.format("%s%02d:|cff%s%s|r", levelcolor, level, colors[class:upper()] or "000000", name), "|cffffff00"..note.. " "..officernote.." |cff00ff00("..rank..")")
			end
		end
	else
		GameTooltip:AddLine("GuildBlock")
		GameTooltip:AddLine(L["Not in a guild"])
	end

	GameTooltip:Show()
end


-----------------------------------------
--      Click to open guild panel      --
-----------------------------------------

function dataobj.OnClick()
	if FriendsFrame:IsVisible() then HideUIPanel(FriendsFrame)
	else
		ToggleFriendsFrame(3)
		FriendsFrame_Update()
		GameTooltip:Hide()
	end
end


-----------------------------------
--      Make rocket go now!      --
-----------------------------------

if IsLoggedIn() then f:PLAYER_LOGIN() else f:RegisterEvent("PLAYER_LOGIN") end


tekBlock/tekBlock.lua

local MINOR = 2
local lib = LibStub:NewLibrary("tekBlock", MINOR)
if not lib then return end


lib.init = lib.init or 0
function lib:new(dataobjname, db)
	-- Delayed init, don't init until first call, and only upgrade that which needs it if previously init'd
	if self.init < 1 then
		self.backdrop = {
			bgFile = "Interface\\Tooltips\\UI-Tooltip-Background",
			edgeFile = "Interface\\Tooltips\\UI-Tooltip-Border",
			edgeSize = 16,
			insets = {left = 5, right = 5, top = 5, bottom = 5},
			tile = true, tileSize = 16,
		}


		function self:OnDragStart()
			if self.db.locked then return end
			local OnLeave = self:GetScript("OnLeave")
			if OnLeave then OnLeave(self) end
			self.OnEnter = self:GetScript("OnEnter")
			self:SetScript("OnEnter", nil)
			self:StartMoving()
		end


		function self:OnDragStop()
			if self.db.locked then return end
			self:SetScript("OnEnter", self.OnEnter)
			if self.OnEnter then self.OnEnter(self) end
			self.OnEnter = nil
			self:StopMovingOrSizing()
			self.db.x, self.db.y = self:GetCenter()
		end


		function self:SetDObjScript(event, name, key, value) self:SetScript(key, value) end
	end

	if self.init < 2 then
		function self:TextUpdate(event, name, key, value)
			self.text:SetText(value)
			if self.db.resize then self:SetWidth(self.text:GetStringWidth() + 8 + (dataobj.icon and 24 or 0)) end
		end

		function self:IconUpdate(event, name, key, value)
			local oldtexture = self.icon:GetTexture()
			self.icon:SetTexture(value)
			if not oldtexture then
				self:SetWidth(self:GetWidth() + 24)
				self.text:SetPoint("CENTER", 12, 0)
			elseif not value then
				self:SetWidth(self:GetWidth() - 24)
				self.text:SetPoint("CENTER", 0, 0)
			end
		end
	end

	self.init = MINOR


	-- And now the good bits
	local ldb = LibStub:GetLibrary("LibDataBroker-1.1")
	local dataobj = ldb:GetDataObjectByName(dataobjname)

	local frame = CreateFrame("Button", nil, UIParent)
	frame:SetHeight(24)

	frame.db = db
	frame:SetPoint("CENTER", UIParent, db.x and "BOTTOMLEFT" or "TOP", db.x or 0, db.y or -100)

	frame:SetMovable(true)
	frame:RegisterForDrag("LeftButton")
	frame:SetClampedToScreen(true)

	frame:SetBackdrop(self.backdrop)
	frame:SetBackdropColor(0.09, 0.09, 0.19, 0.5)
	frame:SetBackdropBorderColor(0.5, 0.5, 0.5, 0.5)

	frame:SetScript("OnDragStart", self.OnDragStart)
	frame:SetScript("OnDragStop", self.OnDragStop)

	frame.icon = frame:CreateTexture()
	frame.icon:SetWidth(16) frame.icon:SetHeight(16)
	frame.icon:SetPoint("LEFT", 8, 0)
	frame.icon:SetTexture(dataobj.icon)
	frame.IconUpdate = self.IconUpdate
	ldb.RegisterCallback(frame, "LibDataBroker_AttributeChanged_"..dataobjname.."_icon", "IconUpdate")

	frame.text = frame:CreateFontString(nil, nil, "GameFontNormalSmall")
	frame.text:SetPoint("CENTER", dataobj.icon and 12 or 0, 0)
	frame.text:SetText(dataobj.text or dataobjname)
	frame:SetWidth(frame.text:GetStringWidth() + 8 + (dataobj.icon and 24 or 0))
	frame.TextUpdate = self.TextUpdate
	ldb.RegisterCallback(frame, "LibDataBroker_AttributeChanged_"..dataobjname.."_text", "TextUpdate")

	frame.SetDObjScript = self.SetDObjScript
	frame:SetScript("OnEnter", dataobj.OnEnter)
	ldb.RegisterCallback(frame, "LibDataBroker_AttributeChanged_"..dataobjname.."_OnEnter", "SetDObjScript")

	frame:SetScript("OnLeave", dataobj.OnLeave)
	ldb.RegisterCallback(frame, "LibDataBroker_AttributeChanged_"..dataobjname.."_OnLeave", "SetDObjScript")

	frame:EnableMouse(true)
	frame:RegisterForClicks("AnyUp")
	frame:SetScript("OnClick", dataobj.OnClick)
	ldb.RegisterCallback(frame, "LibDataBroker_AttributeChanged_"..dataobjname.."_OnClick", "SetDObjScript")

	return frame
end


picoGuild.lua

----------------------------
--      Localization      --
----------------------------

local L = {
	offline = "(.+) has gone offline.";
	online = "|Hplayer:%s|h[%s]|h has come online.";	["has come online"] = "has come online",
	["has gone offline"] = "has gone offline",

	["No Guild"] = "No Guild",
	["Not in a guild"] = "Not in a guild",
}


------------------------------
--      Are you local?      --
------------------------------

local myname = UnitName("player")
local mejoin = myname.." has joined the guild."
local friends, colors = {}, {}
for class,color in pairs(RAID_CLASS_COLORS) do colors[class] = string.format("%02x%02x%02x", color.r*255, color.g*255, color.b*255) end


-------------------------------------------
--      Namespace and all that shit      --
-------------------------------------------

local dataobj = LibStub:GetLibrary("LibDataBroker-1.1"):NewDataObject("picoGuild", {type = "data source", icon = "Interface\\Addons\\picoGuild\\icon", text = L["No Guild"]})
local f = CreateFrame("Frame")
f:SetScript("OnEvent", function(self, event, ...) if self[event] then return self[event](self, event, ...) end end)


----------------------------------
--      Server query timer      --
----------------------------------

local MINDELAY, DELAY = 15, 300
local elapsed, dirty = 0, false
f:Hide()
f:SetScript("OnUpdate", function(self, elap)
	elapsed = elapsed + elap
	if (dirty and elapsed >= MINDELAY) or elapsed >= DELAY then
		if IsInGuild() then GuildRoster() else elapsed, dirty = 0, false end
	end
end)


local orig = GuildRoster
GuildRoster = function(...)
	elapsed, dirty = 0, false
	return orig(...)
end


---------------------------
--      Init/Enable      --
---------------------------

function f:PLAYER_LOGIN()
	LibStub("tekKonfig-AboutPanel").new(nil, "picoGuild")

	self:Show()
	self:RegisterEvent("GUILD_ROSTER_UPDATE")
	self:RegisterEvent("CHAT_MSG_SYSTEM")

	SortGuildRoster("class")
	if IsInGuild() then
		GuildRoster()
	end

	self:UnregisterEvent("PLAYER_LOGIN")
	self.PLAYER_LOGIN = nil
end


------------------------------
--      Event Handlers      --
------------------------------

function f:CHAT_MSG_SYSTEM(event, msg)
	if string.find(msg, L["has come online"]) or string.find(msg, L["has gone offline"]) or msg == mejoin then dirty = true end
end


function f:GUILD_ROSTER_UPDATE()
	if IsInGuild() then
	local total, online = GetNumGuildMembers()
		dataobj.text = string.format("%d/%d", online, total)
	else dataobj.text = L["No Guild"] end
end


------------------------
--      Tooltip!      --
------------------------

local tip = LibStub("tektip-1.0").new(6, "LEFT", "LEFT", "CENTER", "RIGHT", "RIGHT", "RIGHT")
local lastanchor
function dataobj.OnLeave() tip:Hide() end
function dataobj.OnEnter(self)
	tip:AnchorTo(self)
	lastanchor = self

	tip:AddLine("picoGuild")

	if IsInGuild() then

		local gender = UnitSex("player")
		local name, description, standingID, barMin, barMax, barValue = GetGuildFactionInfo()
		local factionStandingtext = GetText("FACTION_STANDING_LABEL"..standingID, gender)
		barMax, barValue = barMax - barMin, barValue - barMin

		tip:AddLine("<"..GetGuildInfo("player")..">", 1, 1, 1)
		tip:AddLine(GetGuildRosterMOTD(), 0, 1, 0, true)
		tip:AddLine(" ")

		tip:AddLine(string.format("Rep:|cffffffff %s %d%% (%d/%d)", factionStandingtext, barValue / barMax * 100, barValue, barMax))
		tip:AddLine(" ")

		local mylevel, myarea = UnitLevel("player"), GetRealZoneText()
		for i=1,GetNumGuildMembers(true) do
			local name, rank, rankIndex, level, class, area, note, officernote, connected, status, engclass = GetGuildRosterInfo(i)
			if connected then
				local cc = RAID_CLASS_COLORS[engclass]
				local lr, lg, lb, ar, ag, ab = 0, 1, 0, 1, 1, 1
				if level < (mylevel - 5) then lr, lg, lb = .6, .6, .6
				elseif level > (mylevel + 5) then lr, lg, lb = 1, 0, 0 end
				if area == myarea then ar, ag, ab = 0, 1, 0 end
				local levelcolor = (level >= (mylevel - 5) and level <= (mylevel + 5)) and "|cff00ff00" or ""
				local grouped = name ~= myname and (UnitInRaid(name) and " |cffffff00[R]" or UnitInParty(name) and " |cffffff00[P]") or ""
				tip:AddMultiLine((level < 10 and "0" or "")..level, name..grouped, area or "???", note, officernote, rank,
					lr,lg,lb, cc.r,cc.g,cc.b, ar,ag,ab, nil,nil,nil, 1,1,0, .7,.7,1)
			end
		end
	else
		tip:AddLine(L["Not in a guild"])
	end

	tip:Show()
end


-----------------------------------------
--      Click to open guild panel      --
-----------------------------------------

function dataobj.OnClick()
	ToggleGuildFrame()
	if GuildFrame:IsShown() then tip:Hide() else dataobj.OnEnter(lastanchor) end
end


-----------------------------------
--      Make rocket go now!      --
-----------------------------------

if IsLoggedIn() then f:PLAYER_LOGIN() else f:RegisterEvent("PLAYER_LOGIN") end
